linkerd/stack/src/fallback.rs

```diff
@@ -59,10 +59,21 @@ impl<B> FallbackLayer<B> {
     where
         E: std::error::Error + 'static,
     {
-        self.with_predicate(|e| e.is::<E>() || e.source().map(|s| s.is::<E>()).unwrap_or(false))
+        self.with_predicate(|e| is_error::<E>(e.as_ref()))
     }
 }
 
+fn is_error<E>(err: &(dyn std::error::Error + 'static)) -> bool
+where
+    E: std::error::Error + 'static,
+{
+    if err.is::<E>() {
+        return true;
+    }
+
+    err.source().map(is_error::<E>).unwrap_or(false)
```

Comment:
when [`error::Chain`](https://doc.rust-lang.org/nightly/std/error/struct.Chain.html) stabilizes, we might just want to use that, but this looks right for now!